### ISSUE TITLE
cuda memcpy instead of gpu kernel in sendrecv operation

### ISSUE THREAD

**Issue by vvmex** [NVIDIA/nccl#1445]
hi, In some cases, the performance of CUDA Memcpy is better than that of GPU kernel.
 is there any way to avoid gpu kernel but cuda memcpy in  sendrecv operation? 
Looking forward to your answer, thanks very much!

**Comment by sjeaugey** [comment]
No, at this time, there isn't a way to completely replace the NCCL operation by a CUDA memcpy. For that, we'd need user buffers to be registered through ncclCommRegister, and then we could consider replacing the NCCL kernel by cudaMemcpy completely, but for now it's not implemented.

**Comment by AVA022** [comment]
> hi, In some cases, the performance of CUDA Memcpy is better than that of GPU kernel. is there any way to avoid gpu kernel but cuda memcpy in p2p sendrecv operation? by the way, NCCL_P2P_USE_CUDA_MEMCPY=1 it will also call the gpu kernel in p2p transport. Looking forward to your answer, thanks very much!

Hi, could you please explain the difference between CUDA Memcpy and GPU kernel as mentioned in your query? I look forward to your response. Thank you!

**Comment by vvmex** [comment]
> No, at this time, there isn't a way to completely replace the NCCL operation by a CUDA memcpy. For that, we'd need user buffers to be registered through ncclCommRegister, and then we could consider replacing the NCCL kernel by cudaMemcpy completely, but for now it's not implemented.

In fact, when using nvswitch, P2P copy bw is better than P2P sm。 At present, I just want to use P2P instead of sendrecv operation. Will this solution be considered?

**Comment by chenhongyu2048** [comment]
as far as I know, p2p cudaMemCopy is not supported in multi-node environment. so this could cause some potential bugs...? I guess

**Comment by adofirst2018** [comment]
> No, at this time, there isn't a way to completely replace the NCCL operation by a CUDA memcpy. For that, we'd need user buffers to be registered through ncclCommRegister, and then we could consider replacing the NCCL kernel by cudaMemcpy completely, but for now it's not implemented.

why cuda memcpy is faster than GPU kernel load/store?

**Comment by AddyLaddy** [comment]
> why cuda memcpy is faster than GPU kernel load/store?
One architectural reason is that the Copy Engines use 256-byte payload transactions whereas SM kernels are limited to 128-bytes.
We have experimented with using the Copy Engines for NCCL collectives, but the extra overheads and complexity of using them made it not worthwhile.

**Comment by adofirst2018** [comment]
> > why cuda memcpy is faster than GPU kernel load/store?
> 
> One architectural reason is that the Copy Engines use 256-byte payload transactions whereas SM kernels are limited to 128-bytes. We have experimented with using the Copy Engines for NCCL collectives, but the extra overheads and complexity of using them made it not worthwhile.

thanks for your reply

**Comment by vvmex** [comment]
> > why cuda memcpy is faster than GPU kernel load/store?
> 
> One architectural reason is that the Copy Engines use 256-byte payload transactions whereas SM kernels are limited to 128-bytes. We have experimented with using the Copy Engines for NCCL collectives, but the extra overheads and complexity of using them made it not worthwhile.

Is there any code for reference, such as tag or pr，thanks very much.

**Comment by sjeaugey** [comment]
Using cudaMemcpy would only work for operations which only do data copy, e.g. allgather or broadcast.

**Comment by Huoyuan100861** [comment]
> Using cudaMemcpy would only work for operations which only do data copy, e.g. allgather or broadcast.

Thank you for your response. I'm wondering if it's possible to entirely avoid launching kernel for allgather or broadcast operations when using copy engine?

**Comment by nullcoderptr** [comment]
> > Using cudaMemcpy would only work for operations which only do data copy, e.g. allgather or broadcast.
> 
> Thank you for your response. I'm wondering if it's possible to entirely avoid launching kernel for allgather or broadcast operations when using copy engine?

You need to implement a CPU function that manages cudamemcpy manually, instead of using NCCL.

**Comment by gbxu** [comment]
> > Using cudaMemcpy would only work for operations which only do data copy, e.g. allgather or broadcast.
> 
> Thank you for your response. I'm wondering if it's possible to entirely avoid launching kernel for allgather or broadcast operations when using copy engine?

You need to implement a CPU function that manages cudamemcpy manually, instead of using NCCL.

**Comment by cll24** [comment]
> > why cuda memcpy is faster than GPU kernel load/store?
> 
> One architectural reason is that the Copy Engines use 256-byte payload transactions whereas SM kernels are limited to 128-bytes. We have experimented with using the Copy Engines for NCCL collectives, but the extra overheads and complexity of using them made it not worthwhile.

Hi, would you please give some more detail about "the extra overheads and complexity" of using Copy Engines. It's a good choice to leverage CE other than SM kernels in terms of performance.

**Comment by sjeaugey** [comment]
Using CE for copy-based operations (in particular for send/recv, alltoall and allgather) is something we have always considered, and we're working on.

Every CE operation has a significant overhead (e.g. 5us) so it wasn't practical for ring algorithms.

But we're working on enabling the use of CE operations with alltoall-style algorithms, where they would make much more sense and provide full offload (and slightly higher peak BW).

**Comment by gbxu** [comment]
> Using CE for copy-based operations (in particular for send/recv, alltoall and allgather) is something we have always considered, and we're working on.
> 
> Every CE operation has a significant overhead (e.g. 5us) so it wasn't practical for ring algorithms.
> 
> But we're working on enabling the use of CE operations with alltoall-style algorithms, where they would make much more sense and provide full offload (and slightly higher peak BW).

If the `cudamemcpyasync` is called in the proxy thread after the AllGather kernel is started, I am worried that it will deadlock with the computation stream using mem* operations.
```
proxy:                                   |->         cudamemcpyasync (blocked by memset)
comm stream:   allgather -|
comp stream:                          memset(blocked by allgather kernel)
```

**Comment by cywork121** [comment]
Hello，Copy Engine performance is poor when copying multiple small message(less than 2MB), and does not reach the total bandwidth of nvlink.. Is there any way to improve it?

**Comment by Zhangmj0621** [comment]
> Hello，Copy Engine performance is poor when copying multiple small message(less than 2MB), and does not reach the total bandwidth of nvlink.. Is there any way to improve it? [@sjeaugey](https://github.com/sjeaugey)
> 
> ![Image](https://github.com/user-attachments/assets/f5a86f1d-a313-4268-9265-3e75f9ca4389)

In fact, if you use SM to do p2p sendrecv opeartion, you still have the trouble to reach total bandwidth of nvlink in small dataSize.
